base/number.jl

```diff
@@ -314,9 +319,12 @@ zero(::Type{Union{}}, slurp...) = Union{}(0)
     one(T::type)
 
 Return a multiplicative identity for `x`: a value such that
-`one(x)*x == x*one(x) == x`.  Alternatively `one(T)` can
-take a type `T`, in which case `one` returns a multiplicative
-identity for any `x` of type `T`.
+`one(x)*x == x*one(x) == x`. If the multiplicative identity can
+be known based on the type alone, then a type may be given as
+an argument to `one` (e.g. `one(Int)` will work because the
+multiplicative identity is the same for all instances of `Int`,
+but `one(Vector{Int})` is not defined because vectors of
+different lengths have different multiplicative identities.)
```

Comment:
I think that is either subtly wrong, or more complicated. The question is which product we're talking about; for a Hadamard product, the multiplicative identity is unique, namely it's `one(Int)` again for any length. We don't define `*(::Vector, ::Vector)`, but if we'd define it as a wedge product, we could in theory get an object that behaves as an identity when plucked into that wedge product. That identity won't be unique and still won't work for any vector length though.

So for `one`, I think we'd either need a different example, or a different justification, since depending on which product `*` is (which depends on the types put in), you either get a unique identity or you don't.

Reply:
Oh right, yeah that was a bad example. Should do `Matrix{Int}` or something like that I guess. If you suggest a change I can accept it, otherwise I'll try and fix it later when I’m home.

Reply:
Alternatively, we could just undo that commit and focus only on `zero`

Reply:
The weird thing is that neither `one` nor `zero` are required to return an object of the same type as the type of the input object (say, for dimensionful quantities where the multiplicative identities are dimensionless), so in theory you can always define an "identity multiplication object" that, on multiplication, just returns the other argument.

This is getting more complicated than we thought on slack :joy:

Reply:
E.g. something like this is always a valid fallback:

```
struct UniversalAdditiveIdentity end
const UAI = UniversalAdditiveIdentity
Base.:+(::UAI, other) = other
Base.:+(other, ::UAI) = other

struct UniversalMultiplicativeIdentity end
const UMI = UniversalMultiplicativeIdentity
Base.:+(::UMI, other) = other
Base.:+(other, ::UMI) = other
```

Reply:
I think `Matrix{Int}` would be a fine example to give. It's allowed to give ab object of a different type, but only in cases where something of the same type makes no sense (e.g. dimensionful quantities)

Reply:
`Matrix` is also weird though, because square matrices do have a multiplicative identity! It's non-square ones that mess things up here entirely.

Reply:
I’m not talking about matrix values, I mean the type

Reply:
```julia

julia> one(rand(2,2))
2×2 Matrix{Float64}:
 1.0  0.0
 0.0  1.0

julia> one(rand(2,2) |> typeof)
ERROR: MethodError: no method matching one(::Type{Matrix{Float64}})

Reply:
Right, but that also has `LinearAlgebra.I` as its identity:

```
julia> I*a
3×2 Matrix{Int64}:
 1  2
 3  4
 5  6
```

I guess `I` doesn't have the shape (technically), but `one(Matrix{Int})` can just return `LinearAlgebra.I`, since it works for any shape :thinking:

Reply:
It's not the same type though.

Reply:
Yes, but it's not required to be. If the same type is actually required, there is `oneunit`, which truly can't be defined for `Matrix{Int}` because of these size/shape issues.